### PR TITLE
SAK-48238 Kernel support single instance store in ContentHosting

### PR DESCRIPTION
**Pull request by csev**

### CODE REVIEW

**Review by ottenhoff — APPROVED** [review]

**Comment by csev** [comment]
@ottenhoff - I kept going and built an MVP for single instance store.

File: kernel/kernel-impl/src/main/sql/mysql/sakai_content.sql [hunk 1]
```diff
@@ -126,6 +126,7 @@ CREATE TABLE CONTENT_RESOURCE
 (
     RESOURCE_ID VARCHAR (255) NOT NULL,
     RESOURCE_UUID VARCHAR (36),
+    RESOURCE_SHA256 VARCHAR (255),
```
Comment (ottenhoff) [inline kernel/kernel-impl/src/main/sql/mysql/sakai_content.sql:129]: please use min chars you need .... indexes get big fast

Comment (csev) [reply kernel/kernel-impl/src/main/sql/mysql/sakai_content.sql:129]: I think 64 would always work - I just don't want to have something blow up.   When you are sure 64 will work - do you use "64" or "99"?

Comment (ottenhoff) [reply kernel/kernel-impl/src/main/sql/mysql/sakai_content.sql:129]: char(64) or varchar(64)

Comment (csev) [reply kernel/kernel-impl/src/main/sql/mysql/sakai_content.sql:129]: Thanks - I will use 64.

File: kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/DbContentService.java [hunk 2]
```diff
@@ -2245,21 +2312,52 @@ private boolean putResourceBodyFilesystem(ContentResourceEdit resource, InputStr
...
+                props.addProperty(ResourceProperties.PROP_CONTENT_SHA256, hex);
                 if (resource.getContentType() != null)
                 {
                     props.addProperty(ResourceProperties.PROP_CONTENT_TYPE, resource.getContentType());
                 }
+
+                // Check if there already is an identical file (most recent if there is > 1)
+                boolean singleInstanceStore = m_serverConfigurationService.getBoolean(PROP_SINGLE_INSTANCE, PROP_SINGLE_INSTANCE_DEFAULT);
+                if ( singleInstanceStore && m_bodyPath != null && m_bodyPath.equals(rootFolder)) {
+                    String statement = "SELECT FILE_PATH FROM "+m_resourceTableName+" WHERE RESOURCE_SHA256 = ? ORDER BY FILE_PATH DESC LIMIT 1;";
```
Comment (jesusmmp) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/DbContentService.java:2337]: New BLOCKER: https://sakaiproject.atlassian.net/browse/SAK-51007

**Pull request merged**